### PR TITLE
FIX #103 - Implementing discovered servers API

### LINKED ISSUE #103: Support Servers and DiscoveredServers APIs
See https://github.com/nats-io/go-nats/issues/214

### PR DESCRIPTION
**Pull request by aricart**
Extended nats_server_control to allow easy creation of a cluster
Bumped versions

### CODE REVIEW

File: lib/nats.js [hunk 1]
```diff
@@ -797,6 +801,23 @@ Client.prototype.processInbound = function() {
 	if (client.checkTLSMismatch() === true) {
 	  return;
 	}
+
+  // Always try to read the connect_urls from info
+  if(client.info.connect_urls && client.info.connect_urls.length > 0) {
```
Comment (kozlovic) [inline lib/nats.js]: First, indentation looks strange.
Second, are you not going to do that for any protocol message received by the client? Looks to me that you should ensure that protocol being processed is an INFO here.
Also, not sure if you support randomization, but what the other clients do is randomize (if allowed) the array or connect_urls prior to add the URLs to the server pool.

File: lib/nats.js [hunk 2]
```diff
@@ -797,6 +801,23 @@ Client.prototype.processInbound = function() {
...
+    // don't add duplicates
+    var known = [];
+    client.servers.forEach(function(server) {
+      known.push(server.url.href);
+    });
+    // add new ones
+    client.info.connect_urls.forEach(function(server) {
+      var u = 'nats://' + server;
+      if(known.indexOf(u) === -1) {
+        client.addServer(u);
```
Comment (derekcollison) [inline lib/nats.js]: Does this do the right thing with randomizing or not the list?

File: test/dyncluster.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+              should(nc.servers.length).be.equal(3);
+              done();
+            }, 1000);
+          });
+        });
+      });
+    });
+  });
+
+  it('setting protocol zero doesnt update', function(done) {
```
Comment (derekcollison) [inline test/dyncluster.js]: Why do we want to expose and let client decide? Is that how we do it in other clients, they can suppress if they want to?

File: lib/nats.js (hunk 2 above)
Comment (kozlovic) [inline lib/nats.js]: Just to be clear, we don't want to randomize the pool when we add things after the initial connect. We can randomize the received URLs *before* adding to the pool, but not randomize the entire pool after initial connect.

Comment (aricart) [inline lib/nats.js]: Yes it does, it is calling addServer which handles randomization options:
https://sourcegraph.com/github.com/nats-io/node-nats@b5db73ee06428471568a730c3b0ff7f907fa6943/-/blob/lib/nats.js#L990:18

File: test/dyncluster.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    servers = nsc.start_cluster([port], route_port, function() {
+      should(servers.length).be.equal(1);
+
+      // connect nats - protocol is set to zero to prevent the info updates
+      var nc = NATS.connect({'port': port, 'reconnectTimeWait': 100, 'protocol': 0});
+      nc.on('connect', function () {
+        // add two servers
+        process.nextTick(function () {
+          var others = nsc.add_member_with_delay([port + 1, port + 2], route_port, 250, function () {
+            should(others.length).be.equal(2);
```
Comment (derekcollison) [inline test/dyncluster.js]: We should write test to check randomize etc. And not randomize.

File: lib/nats.js (hunk 1 above)
Comment (aricart) [inline lib/nats.js]: I preserved the indentation that was in the file :)
The block we are in is only handling INFO messages.  See https://sourcegraph.com/github.com/nats-io/node-nats@b5db73ee06428471568a730c3b0ff7f907fa6943/-/blob/lib/nats.js#L798:1

File: test/dyncluster.js (hunk 3 above)
Comment (aricart) [inline test/dyncluster.js]: I can remove the option.

File: lib/nats.js (hunk 1 above)
Comment (kozlovic) [inline lib/nats.js]: Even with this link, I am sorry but it looks horrible. It is very difficult to know what is in which block.

**Comment by aricart** [comment]
@kozlovic @derekcollison I have put in the changes. `protocol` option is removed. Shuffling of the discovered servers is done and then appended to the current list and an associated test.

**Comment by derekcollison** [comment]
LGTM, let's have @kozlovic give a +1 as well.

File: lib/nats.js [hunk 5]
```diff
@@ -797,6 +797,31 @@ Client.prototype.processInbound = function() {
...
+    var toAdd = [];
+    client.info.connect_urls.forEach(function(server) {
+      var u = 'nats://' + server;
+      if(known.indexOf(u) === -1) {
+        toAdd.push(new Server(url.parse(u)));
+      }
+    });
+
+    if(toAdd.length > 0) {
+      shuffle(toAdd);
```
Comment (kozlovic) [inline lib/nats.js]: If I am not mistaken, this client has a `noRandomize` option. If so, you should shuffle the incoming array only if `noRandomize === false`.

**Comment by aricart** [comment]
@kozlovic I made the shuffling of the discovered servers honor noRandomize, and added a related test.

**Comment by kozlovic** [comment]
LGTM. Not sure why we get so many coveralls report (I deleted some from this page).

**Comment by aricart** [comment]
There were some git errors during checkout for one of the builds.

/a


On May 5, 2017 at 1:15:27 PM, Ivan Kozlovic (notifications@github.com)
wrote:

LGTM. Not sure why we get so many coveralls report (I deleted some from
this page).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/nats-io/node-nats/pull/143#issuecomment-299565082>, or mute
the thread

**Pull request merged**